File: gateway/src/chat/sessions.ts
import type postgres from "postgres";
import { SESSION_CHANNEL } from "../schema.js";

type Sql = ReturnType<typeof postgres>;

export type ChatSession = {
  id: string;
  userId: string;
  personaId: string;
  endedAt: Date | null;
};

async function insertSession(
  sql: Sql,
  userId: string,
  personaId: string,
  channel: (typeof SESSION_CHANNEL)[keyof typeof SESSION_CHANNEL],
): Promise<ChatSession> {
  const rows = await sql<
    {
      id: string;
      user_id: string;
      persona_id: string;
      ended_at: Date | null;
    }[]
  >`
    INSERT INTO sessions (user_id, persona_id, channel)
    VALUES (${userId}, ${personaId}, ${channel})
    RETURNING id, user_id, persona_id, ended_at
  `;
  const row = rows[0];
  if (!row) {
    throw new Error("session insert returned no row");
  }
  return {
    id: row.id,
    userId: row.user_id,
    personaId: row.persona_id,
    endedAt: row.ended_at,
  };
}

export async function createTextSession(
  sql: Sql,
  userId: string,
  personaId: string,
): Promise<ChatSession> {
  return insertSession(sql, userId, personaId, SESSION_CHANNEL.TEXT);
}

export async function createVoiceSession(
  sql: Sql,
  userId: string,
  personaId: string,
): Promise<ChatSession> {
  return insertSession(sql, userId, personaId, SESSION_CHANNEL.VOICE);
}

export async function getSessionForUser(
  sql: Sql,
  sessionId: string,
  userId: string,
): Promise<ChatSession | null> {
  const rows = await sql<
    {
      id: string;
      user_id: string;
      persona_id: string;
      ended_at: Date | null;
    }[]
  >`
    SELECT id, user_id, persona_id, ended_at
    FROM sessions
    WHERE id = ${sessionId}
  `;
  const row = rows[0];
  if (!row) {
    return null;
  }
  if (row.user_id !== userId) {
    return null;
  }
  return {
    id: row.id,
    userId: row.user_id,
    personaId: row.persona_id,
    endedAt: row.ended_at,
  };
}

File: worker/src/worker/persistence/sessions.py
from __future__ import annotations

from typing import Any
from uuid import UUID

import psycopg

SessionRow = dict[str, Any]


def get_session(
    conn: psycopg.Connection,
    session_id: UUID,
    *,
    for_update: bool = False,
) -> SessionRow | None:
    query = """
        SELECT id, user_id, persona_id, engram_session_id, channel,
               started_at, ended_at
        FROM sessions
        WHERE id = %s
    """
    if for_update:
        query = f"{query} FOR UPDATE"
    with conn.cursor() as cur:
        cur.execute(query, (str(session_id),))
        return cur.fetchone()


def create_session(
    conn: psycopg.Connection,
    *,
    user_id: UUID,
    persona_id: UUID,
    channel: str,
) -> SessionRow:
    with conn.cursor() as cur:
        cur.execute(
            """
            INSERT INTO sessions (user_id, persona_id, channel)
            VALUES (%s, %s, %s)
            RETURNING id, user_id, persona_id, engram_session_id, channel,
                      started_at, ended_at
            """,
            (str(user_id), str(persona_id), channel),
        )
        row = cur.fetchone()
    if row is None:
        raise RuntimeError("session insert returned no row")
    return row


def set_engram_session_id(
    conn: psycopg.Connection,
    session_id: UUID,
    engram_session_id: str,
) -> None:
    with conn.cursor() as cur:
        cur.execute(
            """
            UPDATE sessions
            SET engram_session_id = %s
            WHERE id = %s
            """,
            (engram_session_id, str(session_id)),
        )


def user_engram_id(conn: psycopg.Connection, user_id: UUID) -> str | None:
    with conn.cursor() as cur:
        cur.execute(
            "SELECT engram_user_id FROM users WHERE id = %s",
            (str(user_id),),
        )
        row = cur.fetchone()
    if row is None:
        return None
    value = row["engram_user_id"]
    return value if isinstance(value, str) else None


def user_email(conn: psycopg.Connection, user_id: UUID) -> str | None:
    with conn.cursor() as cur:
        cur.execute("SELECT email FROM users WHERE id = %s", (str(user_id),))
        row = cur.fetchone()
    if row is None:
        return None
    value = row["email"]
    return value if isinstance(value, str) else None
